bindings/rust/bench/src/s2n_tls.rs

```diff
@@ -0,0 +1,206 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    harness::{read_to_bytes, Mode, TlsBenchHarness},
+    CA_CERT_PATH, SERVER_CERT_CHAIN_PATH, SERVER_KEY_PATH,
+};
+use log::debug;
+use s2n_tls::{
+    callbacks::VerifyHostNameCallback,
+    config::{Builder, Config},
+    connection::Connection,
+    enums::Blinding,
+    security::DEFAULT_TLS13,
+};
+use std::{
+    collections::VecDeque,
+    ffi::c_void,
+    io::{Read, Write},
+    os::raw::c_int,
+    pin::Pin,
+    task::Poll::Ready,
+};
+
+pub struct S2nTls {
+    client_to_server_buf: Pin<Box<VecDeque<u8>>>, // Pinned pointer to a VecDeque to pass as C pointer for custom IO
+    server_to_client_buf: Pin<Box<VecDeque<u8>>>,
+    client_config: Config,
+    server_config: Config,
+    client_conn: Connection,
+    server_conn: Connection,
+    client_handshaked: bool,
+    server_handshaked: bool,
+}
+
+/// Custom callback for verifying hostnames, need it to use s2n-tls safely
+struct HostNameHandler<'a> {
+    expected_server_name: &'a str,
+}
+impl VerifyHostNameCallback for HostNameHandler<'_> {
+    fn verify_host_name(&self, hostname: &str) -> bool {
+        self.expected_server_name == hostname
+    }
+}
+
+impl S2nTls {
+    /// Unsafe callback for custom IO C API
+    unsafe extern "C" fn send_cb(context: *mut c_void, data: *const u8, len: u32) -> c_int {
```

Comment:
This is probably just my ignorance with rust showing, but why do you need custom send/recv callbacks?

Reply:
The rust bindings for s2n-tls only expose the custom IO, so we need custom send/recv callbacks and contexts.

Reply:
I think it might be useful to call out the additional motivation of skipping the networking stack. Writing over localhost/writing to a fd would move more things outside of the benchmark harness and add more variables, and therefore would make it harder to directly compare different TLS implementations and get consistent benchmarks.

Reply:
Added comment, hopefully more clear now.